source/common/network/connection_impl.cc

```diff
@@ -888,6 +888,16 @@ ClientConnectionImpl::ClientConnectionImpl(
       stream_info_(dispatcher_.timeSource(), socket_->connectionInfoProviderSharedPtr()) {
 
   stream_info_.setUpstreamInfo(std::make_shared<StreamInfo::UpstreamInfoImpl>());
+
```

Comment:
IIUC, we moving this ahead is because the following code will return for internal connection since its remoteAddress type is EnvoyInternal, not IP?

  if (socket_->connectionInfoProviderSharedPtr()->remoteAddress()->ip() == nullptr) {
    return;
  }

https://github.com/envoyproxy/envoy/blob/bd480aa2a62af2885f27983f6d1fa96cf51e8fdf/source/common/network/connection_impl.cc#L893

https://github.com/envoyproxy/envoy/blob/bd480aa2a62af2885f27983f6d1fa96cf51e8fdf/envoy/network/address.h#L198

However, one thing confusing me is that even without this change, we had observed that the filter state stored by the external listener filter are able to shared to upstream and got retrieved correctly by the filters in the internal listener.

Reply:
Yes, internal listener merge is in a different location here:

https://github.com/envoyproxy/envoy/blob/bd480aa2a62af2885f27983f6d1fa96cf51e8fdf/source/extensions/bootstrap/internal_listener/active_internal_listener.cc#L65

So what happens is that a shared object in the downstream filter state for the internal listener, but not in the upstream connection filter state. From downstream, we can access the upstream filter state but not the next internal downstream filter state.